File: skills/memory_manager.py
import logging
from skills.base_skill import BaseSkill

logger = logging.getLogger(__name__)


class MemoryManagerSkill(BaseSkill):
    def __init__(self, db):
        self._db = db

    @property
    def name(self) -> str:
        return "memory_manager"

    @property
    def description(self) -> str:
        return (
            "Verwalte Claras Langzeitgedaechtnis. Speichere, suche, lese und loesche "
            "Erinnerungen ueber den Nutzer, seine Vorlieben und wichtige Fakten."
        )

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "enum": [
                        "remember", "recall", "recall_category", "search",
                        "forget", "list_categories", "stats",
                    ],
                    "description": (
                        "Aktion: remember (speichern), recall (einzeln abrufen), "
                        "recall_category (Kategorie abrufen), search (suchen), "
                        "forget (loeschen), list_categories (Kategorien auflisten), "
                        "stats (Statistiken)"
                    ),
                },
                "category": {
                    "type": "string",
                    "description": (
                        "Kategorie der Erinnerung, z.B. 'vorlieben', 'persoenlich', "
                        "'technik', 'ziele', 'projekte'"
                    ),
                },
                "key": {
                    "type": "string",
                    "description": "Schluessel/Name der Erinnerung",
                },
                "value": {
                    "type": "string",
                    "description": "Wert/Inhalt der Erinnerung",
                },
                "query": {
                    "type": "string",
                    "description": "Suchbegriff fuer die Suche in Erinnerungen",
                },
            },
            "required": ["action"],
        }

    async def execute(self, **kwargs) -> str:
        action = kwargs.get("action", "")
        category = kwargs.get("category", "")
        key = kwargs.get("key", "")
        value = kwargs.get("value", "")
        query = kwargs.get("query", "")

        if action == "remember":
            if not category or not key or not value:
                return "Fehler: 'category', 'key' und 'value' sind erforderlich."
            await self._db.remember(category, key, value)
            return f"Gespeichert: [{category}] {key} = {value}"

        elif action == "recall":
            if not category or not key:
                return "Fehler: 'category' und 'key' sind erforderlich."
            result = await self._db.recall(category, key)
            if result:
                return f"[{category}] {key} = {result}"
            return f"Keine Erinnerung gefunden fuer [{category}] {key}"

        elif action == "recall_category":
            if not category:
                return "Fehler: 'category' ist erforderlich."
            entries = await self._db.recall_category(category)
            if not entries:
                return f"Keine Erinnerungen in Kategorie '{category}'"
            lines = [f"Kategorie '{category}' ({len(entries)} Eintraege):"]
            for e in entries:
                lines.append(f"  - {e['key']}: {e['value']}")
            return "\n".join(lines)

        elif action == "search":
            if not query:
                return "Fehler: 'query' ist erforderlich."
            results = await self._db.search_memory(query)
            if not results:
                return f"Keine Ergebnisse fuer '{query}'"
            lines = [f"Suchergebnisse fuer '{query}' ({len(results)} Treffer):"]
            for r in results:
                lines.append(f"  - [{r['category']}] {r['key']}: {r['value']}")
            return "\n".join(lines)

        elif action == "forget":
            if not category or not key:
                return "Fehler: 'category' und 'key' sind erforderlich."
            await self._db.forget(category, key)
            return f"Geloescht: [{category}] {key}"

        elif action == "list_categories":
            categories = await self._db.get_all_categories()
            if not categories:
                return "Noch keine Erinnerungen gespeichert."
            return "Kategorien: " + ", ".join(categories)

        elif action == "stats":
            count = await self._db.count_memories()
            categories = await self._db.get_all_categories()
            lines = [
                f"Gedaechtnis-Statistiken:",
                f"  Eintraege gesamt: {count}",
                f"  Kategorien: {len(categories)}",
            ]
            if categories:
                lines.append(f"  Kategorienamen: {', '.join(categories)}")
            return "\n".join(lines)

        return f"Unbekannte Aktion: {action}"

File: src/tools/builtin/memory-manager.ts
import type { Tool } from '../registry.js'
import type { MemoryManager } from '../../memory/manager.js'

export class MemoryManagerTool implements Tool {
  readonly name = 'memory_manager'
  readonly description = 'Verwaltet das Langzeitgedaechtnis: Fakten speichern, abrufen, suchen und loeschen.'
  readonly parameters = {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['remember', 'recall', 'search', 'forget', 'list'],
        description: 'Die auszufuehrende Aktion',
      },
      category: {
        type: 'string',
        description: 'Kategorie: vorlieben, persoenlich, technik, ziele, projekte, gewohnheiten, wichtig',
      },
      key: { type: 'string', description: 'Eindeutiger Bezeichner des Eintrags' },
      value: { type: 'string', description: 'Zu speichernder Wert (nur bei action=remember)' },
      query: { type: 'string', description: 'Suchbegriff (nur bei action=search)' },
    },
    required: ['action'],
  }

  constructor(private readonly memory: MemoryManager) {}

  async execute(args: Record<string, unknown>): Promise<string> {
    const action = String(args.action ?? '')
    const category = String(args.category ?? '')
    const key = String(args.key ?? '')
    const value = String(args.value ?? '')
    const query = String(args.query ?? '')

    switch (action) {
      case 'remember': {
        if (!category || !key || !value) return 'Fehler: category, key und value erforderlich.'
        this.memory.remember(category, key, value)
        return `Gespeichert: [${category}] ${key} = ${value}`
      }
      case 'recall': {
        if (!category || !key) return 'Fehler: category und key erforderlich.'
        const db = (this.memory as unknown as { db: { recall: (c: string, k: string) => string | null } }).db
        const val = db.recall(category, key)
        return val ? `[${category}] ${key}: ${val}` : `Kein Eintrag fuer '${key}' in Kategorie '${category}'.`
      }
      case 'search': {
        if (!query) return 'Fehler: query erforderlich.'
        const results = await this.memory.search(query, 10)
        if (!results.length) return `Keine Ergebnisse fuer '${query}'.`
        return results.map(r => `[${r.category}] ${r.key}: ${r.value}`).join('\n')
      }
      case 'forget': {
        if (!category || !key) return 'Fehler: category und key erforderlich.'
        this.memory.forget(category, key)
        return `Eintrag geloescht: [${category}] ${key}`
      }
      case 'list': {
        const memories = category
          ? (this.memory as unknown as { db: { recallCategory: (c: string) => Array<{ key: string; value: string }> } }).db.recallCategory(category)
            .map(m => `[${category}] ${m.key}: ${m.value}`)
          : this.memory.getRecentMemories(20).map(m => `[${m.category}] ${m.key}: ${m.value}`)
        return memories.length ? memories.join('\n') : 'Keine Eintraege gefunden.'
      }
      default:
        return `Unbekannte Aktion: ${action}`
    }
  }
}
